bindings/c/test/apitester/TesterBlobGranuleCorrectnessWorkload.cpp

```diff
@@ -348,6 +355,199 @@ class ApiBlobGranuleCorrectnessWorkload : public ApiWorkload {
+
+		// TODO have delta mutations update map
+	}
+
+	void validateBGDescriptionData(std::shared_ptr<ITransactionContext> ctx,
+	                               fdb::native::FDBReadBlobGranuleContext& bgCtx,
+	                               fdb::GranuleDescription desc,
+	                               fdb::Key begin,
+	                               fdb::Key end,
+	                               int64_t readVersion) {
+		ASSERT(desc.keyRange.beginKey < desc.keyRange.endKey);
+		// beginVersion of zero means snapshot present
+
+		// validate snapshot file
+		ASSERT(desc.snapshotFile.has_value());
+		if (BG_API_DEBUG_VERBOSE) {
+			info(fmt::format("Loading snapshot file {0}\n", fdb::toCharsRef(desc.snapshotFile->filename)));
+		}
+		validateSnapshotData(ctx, bgCtx, *desc.snapshotFile, desc.keyRange);
+
+		// validate delta files
+		int64_t lastDFMaxVersion = 0;
+		for (int i = 0; i < desc.deltaFiles.size(); i++) {
+			validateDeltaData(ctx, bgCtx, desc.deltaFiles[i], desc.keyRange, lastDFMaxVersion);
+		}
+
+		// validate memory mutations
+		int64_t lastVersion = 0;
+		for (int i = 0; i < desc.memoryMutations.size(); i++) {
+			fdb::GranuleMutation& m = desc.memoryMutations[i];
+			ASSERT(m.type == 0 || m.type == 1);
+			ASSERT(m.version > 0);
+			ASSERT(m.version >= lastVersion);
+			ASSERT(m.version <= readVersion);
+			lastVersion = m.version;
+
+			ASSERT(m.type == 0 || m.type == 1);
+			ASSERT(desc.keyRange.beginKey <= m.param1);
+			ASSERT(m.param1 < desc.keyRange.endKey);
+			if (m.type == 1) {
+				ASSERT(desc.keyRange.beginKey <= m.param2);
+				ASSERT(m.param2 <= desc.keyRange.endKey);
+			}
+
+			// TODO have delta mutations update map
+		}
+
+		// TODO: validate map against data store
+	}
+
+	void validateBlobGranuleDescriptions(std::shared_ptr<ITransactionContext> ctx,
+	                                     std::vector<fdb::GranuleDescription> results,
+	                                     fdb::Key begin,
+	                                     fdb::Key end,
+	                                     std::optional<int> tenantId,
+	                                     int64_t readVersion) {
+		ASSERT(!results.empty());
+		ASSERT(results.front().keyRange.beginKey <= begin);
+		ASSERT(end <= results.back().keyRange.endKey);
+		for (int i = 0; i < results.size() - 1; i++) {
+			ASSERT(results[i].keyRange.endKey == results[i + 1].keyRange.beginKey);
+		}
+
+		if (tenantId) {
+			// FIXME: support tenants!!
+			info("Skipping validation because of tenant.");
+			return;
+		}
+
+		TesterGranuleContext testerContext(ctx->getBGBasePath());
+		fdb::native::FDBReadBlobGranuleContext bgCtx = createGranuleContext(&testerContext);
+		for (int i = 0; i < results.size(); i++) {
+			validateBGDescriptionData(ctx, bgCtx, results[i], begin, end, readVersion);
+		}
+	}
+
+	void randomReadDescription(TTaskFct cont, std::optional<int> tenantId) {
+		if (!seenReadSuccess(tenantId)) {
+			return;
+		}
+		fdb::Key begin = randomKeyName();
```

Comment:
This is going to hit the same problem as a bug I just hit in a random other PR on main, fixed by https://github.com/apple/foundationdb/pull/9255.
You'd likely want to rebase on this and use randomKeyRange() here instead before merging